Adds command `flow owner remove`

MathijsVerbeeck opened the pull request:
Closes #4560

milanholemans commented:
Awesome work @MathijsVerbeeck! We'll try to review it ASAP!

milanholemans commented on src/m365/flow/commands/owner/owner-remove.ts:
```diff
@@ -0,0 +1,152 @@
+      },
+      {
+        option: '--asAdmin'
+      }
```
Seems like we're missing a confirm flag.

milanholemans commented on src/utils/aadGroup.ts:
```diff
@@ -25,19 +25,21 @@ export const aadGroup = {
    */
-  async getGroupByDisplayName(displayName: string): Promise<Group> {
-    const groups = await this.getGroupsByDisplayName(displayName);
+  async getGroupByDisplayName(displayName: string, selectProperties?: string[]): Promise<Group> {
```
Instead of this, I think we rather create a new function `getGroupIdByDisplayName` like we did for `aadUser`.

milanholemans commented on src/m365/flow/commands/owner/owner-remove.spec.ts:
```diff
@@ -0,0 +1,194 @@
+    sinon.stub(auth, 'restoreAuth').callsFake(() => Promise.resolve());
+    sinon.stub(telemetry, 'trackEvent').callsFake(() => { });
+    sinon.stub(pid, 'getProcessName').callsFake(() => '');
+    auth.service.connected = true;
```
Could you add the extra test stubs?
https://github.com/pnp/cli-microsoft365/pull/4545/files#diff-195628c809e6fb2923588c891e9fe2362a4422873efef5b27532a051cf9d065d

milanholemans commented on src/m365/flow/commands/owner/owner-remove.spec.ts:
```diff
@@ -0,0 +1,194 @@
+    });
+
+    await command.action(logger, { options: { verbose: true, environmentName: environmentName, name: flowName, userName: userName } });
+    assert.deepStrictEqual(JSON.stringify(postStub.lastCall.args[0].data), JSON.stringify(requestBodyUser));
```
Let's remove the stringify.

milanholemans commented on src/m365/flow/commands/owner/owner-remove.spec.ts:
```diff
@@ -0,0 +1,194 @@
+      new CommandError(error.error.message));
+  });
+
+  it('throws error when Flow not found', async () => {
```
I think 1 failing test will suffice.

milanholemans reviewed (changes requested):
Command works great @MathijsVerbeeck!
Let's make some changes before we proceed.

MathijsVerbeeck commented:
Thanks for the review @milanholemans 

Had some issues resolving your requests, but should have fixed them all! 😄

milanholemans commented on docs/docs/cmd/flow/owner/owner-remove.md:
```diff
@@ -0,0 +1,67 @@
+`-e, --environmentName <environmentName>`
+: The name of the environment.
+
+`-n, --flowName <flowName>`
```
Let's use short option `-f`.

milanholemans commented on src/m365/flow/commands/owner/owner-remove.ts:
```diff
@@ -0,0 +1,174 @@
+  groupId?: string;
+  groupName?: string;
+  asAdmin?: boolean;
+}
```
We're missing a confirm option here.

milanholemans commented on src/m365/flow/commands/owner/owner-remove.ts:
```diff
@@ -0,0 +1,174 @@
+        option: '-e, --environmentName <environmentName>'
+      },
+      {
+        option: '-n, --flowName <flowName>'
```
```suggestion
        option: '-f, --flowName <flowName>'
```

milanholemans commented on src/m365/flow/commands/owner/owner-remove.ts:
```diff
@@ -0,0 +1,174 @@
+          type: 'confirm',
+          name: 'continue',
+          default: false,
+          message: `Are you sure you want to remove the Flow owner from the specified flow?`
```
Let's show the owner in this message as well.

milanholemans commented on src/utils/aadGroup.spec.ts:
```diff
@@ -90,4 +90,62 @@ describe('utils/aadGroup', () => {
+
+    try {
+      await aadGroup.getGroupIdByDisplayName(validGroupName);
+      assert.fail('Error expected, but was not thrown.');
```
Let's use `assert.rejects` instead of this construction with try/catches just like we do for regular commands.

milanholemans commented on src/utils/aadGroup.spec.ts:
```diff
@@ -90,4 +90,62 @@ describe('utils/aadGroup', () => {
+
+    try {
+      await aadGroup.getGroupIdByDisplayName(validGroupName);
+      assert.fail('Error expected, but was not thrown.');
```
Let's use `assert.rejects` instead of this construction with try/catches just like we do for regular commands.

milanholemans commented on src/utils/aadGroup.ts:
```diff
@@ -25,6 +25,7 @@ export const aadGroup = {
   /**
    * Get a list of groups by display name.
    * @param displayName Group display name.
+   * @param selectProperties Properties to select if specified.
```
Is there a reason why this description was added?

milanholemans commented on src/utils/aadGroup.ts:
```diff
@@ -33,6 +34,7 @@ export const aadGroup = {
   /**
    * Get a single group by its display name.
    * @param displayName Group display name.
+   * @param selectProperties Properties to select if specified.
```
Is there a reason why this description was added?

MathijsVerbeeck replied:
I used this method before. My bad

milanholemans commented on src/utils/aadGroup.ts:
```diff
@@ -48,5 +50,25 @@ export const aadGroup = {
+   * @throws Error when multiple groups with the same name were found.
+   */
+  async getGroupIdByDisplayName(displayName: string): Promise<string> {
+    const groups = await odata.getAllItems<Group>(`${graphResource}/v1.0/groups?$filter=displayName eq '${formatting.encodeQueryParameter(displayName)}'?$select=id`);
```
When executing this function, I'm getting an error because the URL is invalid.

```suggestion
    const groups = await odata.getAllItems<Group>(`${graphResource}/v1.0/groups?$filter=displayName eq '${formatting.encodeQueryParameter(displayName)}'&$select=id`);
```

milanholemans reviewed (changes requested):
Hi @MathijsVerbeeck 

Nice job so far, I noticed a few things we should change before we merge it. Could you have a look at them please?

MathijsVerbeeck commented:
Hi @milanholemans 

I've implemented your requested changes. Please note that I've also rewritten the other error handling tests in the `aadgroup.spec.ts` file.

Could you please have another look?

milanholemans reviewed (approved):
Looks good to me, ready to merge!

milanholemans commented:
Merged manually, thank you for this great addition!
